core/src/allocation.jl

```diff
@@ -0,0 +1,503 @@
+            if i == 1 || i == length(AG_edge_composite)
+                continue
+            end
+            node_type = lookup[subnetwork_node_id]
+            node = getfield(p, node_type)
+
+            # Find flow constraints
+            if is_flow_constraining(node)
+                model_node_idx = Ribasim.findsorted(node.node_id, subnetwork_node_id)
+                AG_edge_capacity = min(AG_edge_capacity, node.max_flow_rate[model_node_idx])
+            end
+
+            # Find flow direction constraints
+            if is_flow_direction_constraining(node)
+                subnetwork_inneighbor_node_id =
+                    only(inneighbors(graph_flow, subnetwork_node_id))
+
+                if subnetwork_inneighbor_node_id == AG_edge_composite[i - 1]
+                    negative_flow = false
+                elseif subnetwork_inneighbor_node_id == AG_edge_composite[i + 1]
+                    positive_flow = false
+                end
+            end
+        end
+
+        # Add composite AG edge(s)
+        if positive_flow
+            add_edge!(graph_allocation, AG_node_id_1, AG_node_id_2)
+            capacity[AG_node_id_1, AG_node_id_2] = AG_edge_capacity
+        end
+
+        if negative_flow
+            add_edge!(graph_allocation, AG_node_id_2, AG_node_id_1)
+            capacity[AG_node_id_2, AG_node_id_1] = AG_edge_capacity
+        end
+    end
+
+    # The source nodes must only have one outneighbor
+    for (AG_node_id, AG_node_type) in values(node_id_mapping)
+        if AG_node_type == :source
+            if !(
+                (length(inneighbors(graph_allocation, AG_node_id)) == 0) &&
+                (length(outneighbors(graph_allocation, AG_node_id)) == 1)
+            )
+                @error "Sources nodes in the max flow graph must have no inneighbors and 1 outneighbor."
+                errors = true
+            end
+        end
+    end
+
+    # Invert the node id mapping to easily translate from AG nodes to subnetwork nodes
+    node_id_mapping_inverse = Dict{Int, Tuple{Int, Symbol}}()
+
+    for (subnetwork_node_id, (AG_node_id, node_type)) in node_id_mapping
+        node_id_mapping_inverse[AG_node_id] = (subnetwork_node_id, node_type)
+    end
+
+    # Remove user return flow edges that are upstream of the user itself
+    AG_node_ids_user = [
+        AG_node_id for
+        (AG_node_id, node_type) in values(node_id_mapping) if node_type == :user
+    ]
+    AG_node_ids_user_with_returnflow = Int[]
+    for AG_node_id_user in AG_node_ids_user
+        AG_node_id_return_flow = only(outneighbors(graph_allocation, AG_node_id_user))
+        if path_exists(graph_allocation, AG_node_id_return_flow, AG_node_id_user)
+            rem_edge!(graph_allocation, AG_node_id_user, AG_node_id_return_flow)
+            # TODO: Add to logging?
+            @warn "The outflow of user #$(node_id_mapping_inverse[AG_node_id_user][1]) is upstream of this user itself and thus ignored."
```

Comment:
This is a valid situation, so it's more informational, since warnings tend to be for things users should change. Perhaps a debug statement is better?
Also it would be good to explain what is ignored, reading it at face value modelers could think the outflow does not happen.